### DISCUSSION TITLE
A lot left unsaid in examples

### DISCUSSION THREAD

**Discussion by kylemh** [vercel/turborepo#983] [Ideas]
I wanted to open this as an issue to make it actionable, but I just wanted to point out that a LOT is left unsaid in multiple examples which is going to lead to issues or FAQ. Most of the issues are centered on the fact that users are encouraged to transpile across packages to avoid extra build steps. I opened this issue to get some answers before I attempt to improve the examples with some PRs 🙏🏼 

For example:
- Path aliases: [[Source](https://github.com/vercel/turborepo/discussions/620)] - When using aliases in one package's `tsconfig`, you need to also map them to the tsconfig of consuming libraries. The approved answer in this discussion is actually potentially disastrous if there exists a module in the application's filesystem when the user meant to reference a package's filesystem. I tried to solve this by using `no-restricted-imports` lint rule to ban relative imports and encourage devs to use a path that DID NOT exist in consuming apps. For example, if every app had `paths: { '~/*': ['./*'] }` defined, the library will need to do `paths: { '@root': './index.ts' }`. My solution could still fail somebody someday since it's theoretically possible that an application could create a similar alias.
- Storybook and consequences: [[Source](https://github.com/vercel/turborepo/discussions/289)] - A LOT of people are going to want to use Storybook or Styleguidist or something similar with shared component libraries. They're also going to want to put it directly into the package they're documenting for many reasons (file colocation and Storybook's react docgen come to mind quickly. It'd be pretty trivial to set this up in the design system example, but for the other examples where using `next-transpile-modules` is encouraged, introducing Storybook leads to a lot of consequences. For example, at that point, somebody would need to install `React` as a devDep, but this could lead to an application shipping multiple instances of React (as covered [here in `next-transpile-modules` docs](https://github.com/martpie/next-transpile-modules#i-have-trouble-with-duplicated-dependencies-or-the-invalid-hook-call-error-in-react)). [I don't see a perfect solution to this problem](https://github.com/martpie/next-transpile-modules/issues/260) and `next-transpile-modules`'s documentation also acknowledges that their recommended solution isn't perfect. Rather than document, it may be best to change _all_ the examples to have a build step to avoid this issue, but maybe somebody has a better idea or maybe the preference is to encourage transpiling and updating the examples to count for these issues? One last point on Storybook, the dev server has a cache folder which could be used to `turborepo`'s advantage (`node_modules/.cache/storybook`). Doing an example with Storybook could help for that reason alone.
- There's no explanation why - in the pnpm example - packages are [installed via `workspaces:*`](https://github.com/vercel/turborepo/blob/main/examples/with-pnpm/apps/web/package.json#L15) and others are [installed via the package name with the "any version" indicator (*)](https://github.com/vercel/turborepo/blob/main/examples/basic/apps/web/package.json#L15). I think `workspaces:*` is the reason [this](https://github.com/vercel/turborepo/issues/331) occurs because I install "packages" in "apps" with no type acquisition issues.
- I haven't run into it, but I'd also love documentation around avoiding circular imports

**Comment by mattpocock** [comment]
I'm now working on Turborepo docs 3 days a week, I'm making this one of my top priorities.

**Comment by hubertkuoch** [comment]
Since the docs and the examples recommend next-transpile-modules, I think it is important to check its implication when building apps. For instance, just tried to build an internal package on top of @emotion/server which is widely used for enabling styling SSR with Next.js and depending on whether this package is installed only at the internal package level, at the consuming next app level or at the root of a pnpm workspace, the server bundle is not consistent size-wise. I imagine this can also be the case with other third-party packages so it may be worth checking into this.

**Comment by kylemh** [comment]
I imagine [typeRoots being consulted during module resolution](https://devblogs.microsoft.com/typescript/announcing-typescript-5-1/#typeroots-are-consulted-in-module-resolution) can have a positive effect on the first bullet of my original post.

**Comment by anthonyshew** [comment]
Hey, Kyle, thanks for these notes. Really helpful!

TL;DR: **We need to find a balance on creating examples that are a great start for understanding monorepos. Not too little, not too much.** I personally think we're pretty close and we can continue to do better.

Direct answers to your questions at the bottom.

---

My initial reaction on these is that those of us on the Turborepo team want to be mindful of our role in the ecosystem today. Matt did an excellent job with his work on these and I spent the first couple weeks of my time here as Turbo DX working on the examples, too. Here are my thoughts (with emojis on where I personally think we stand right now, today): 
- We **do** want to provide useful examples. ✅ 
- We **do** want the examples to help you feel the power of Turborepo right away. ✅
- We **do** want the Turbo community to enable folks to have a healthy working knowledge of monorepo tooling. We hope that, if someone has monorepo questions, they know they can visit the Turbo community and find answers. ❌ 
- We **do** want folks to know that Turborepo is framework/tooling agnostic. Bring `turbo` to your JavaScript/TypeScript monorepo and it will get faster. ✅ 
- We **do** want to have a set of examples that we can reasonably maintain over time. ✅ 

The last bullet point is really what makes expanding the examples tough - and leads me to the "don't" category:

- We **don't** want to be chasing intricacies of each individual tool. JavaScript ecosystem tooling very quickly adds up and, the deeper you go with one, the more you have to chase the intersection of all of the tools combined. The maintenance burden becomes exponential. ✅ 
- We **don't** want to re-document individual, non-Turborepo tools. We'd like to be active in helping the JavaScript community ensure the tools they build are great for monorepos - but we want to be users of those tools with the rest of the community rather than knowledge keepers. (Over time, we may find that Turborepo users want us to create tooling that overlaps with those other tools, as well.) ✅ 
- We **don't** want the examples to be a burden on our team that detracts from improving Turborepo. Every minute the core team has had to spend finicking with these examples is a minute lost towards making Turborepo awesome. (Certainly an investment in helping folks, though, not to be discounted.) ✅ 
- We **don't** want the examples to be confusing for beginners. ⚠️ 
- We **don't** want the examples/documentation to be missing answers to questions that the community is asking too often. ⚠️ 


---

There are probably more list items I can add to the above context but hopefully it paints a good enough picture of what we're looking for:
- The examples show you the very basics of Turborepo, monorepos, and overwhelmingly common ecosystem tools.
- By nature, the examples **should** leave some things unsaid.
- They showcase Turborepo's flexibility to work with the ecosystem and different ways you can use `turbo` for most common tasks.

Some additional useful context may be that I asked the Next.js team about their [examples directory](https://github.com/vercel/next.js/tree/canary/examples). To make a long story short, they've said it's too big. We're going to piggyback on that experience and try to keep things more lean here in Turbo.

---
# Directly answering your questions to the best of my knowledge finally

- Path aliases: This is indeed a shortcoming of the `transpilePackages` strategy that you'll find in our examples. I personally haven't found a way of mitigating this that's totally bulletproof. It's largely a matter of tradeoffs, as you'll find folks discussing in the Discussion you've mentioned. Some folks may decide that path aliases make it worth bundlers parsing their code twice to produce a build. Some may not. However, this is something that does get asked about frequently and I'm hoping to have a better answer soon.
- Storybook: It makes the most sense to have Storybook be an application in your monorepo for demonstration purposes. In the event that you change Storybook-only source code, you would only want to miss cache for that application. If Storybook is lumped into the `ui` package, you would force the rest of the monorepo to miss cache (even though you only updated a Storybook story.) There are, of course, ways around this (namely, the `inputs` key in `turbo.json`) - but this gets too far into the weeds too quickly for a simple demonstration for beginners. If you're someone who is thinking about these tradeoffs already, you likely don't need our help as much as folks trying to get the basics. 😄
- pnpm: yarn v1/2 use `*` to refer to other workspaces and yarn 3+ and pnpm use `workspace:*`. This is in our documentation today (it may not have been when you asked the question).
- I have it on my todo list to write a circular imports document.

Thanks again for your thoughtfulness here, Kyle!
